### ISSUE TITLE
About using swinv2 for transfer training at high resolution

### ISSUE THREAD

**Issue by CodeStarting-design** [microsoft/Swin-Transformer#335]
Hello, I tried to use an 8*8 window and a resolution of 256*256 to pre-train swinV2, and then used a 16*16 window and a resolution of 512*512 for final training. Can you provide an example code？

**Comment by fosteman** [comment]
something like this could work

`import gc
import time
import json
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, WeightedRandomSampler
from torchvision import transforms, datasets
from transformers import SwinForImageClassification, SwinConfig
from tqdm import tqdm
from sklearn.metrics import classification_report
import numpy as np
from datetime import datetime
import comet_ml


max_workers = 2
prefetch_factor=2
# Set batch size (adjust based on available memory)
batch_size = 8  # Adjust based on your environment

# Verbose flag for detailed output
verbose = True

class FocalLoss(nn.Module):
    def __init__(self, alpha=0.25, gamma=2.0, reduction='mean'):
        super(FocalLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.reduction = reduction

    def forward(self, inputs, targets):
        ce_loss = nn.CrossEntropyLoss(reduction='none')(inputs, targets)
        pt = torch.exp(-ce_loss)  # pt is the probability of correctly classifying the current sample
        focal_loss = self.alpha * (1 - pt) ** self.gamma * ce_loss

        if self.reduction == 'mean':
            return focal_loss.mean()
        elif self.reduction == 'sum':
            return focal_loss.sum()
        else:
            return focal_loss

def clear_memory():
    """Clears memory and cache to avoid memory overflow."""
    gc.collect()
    torch.cuda.empty_cache()



def append_metrics_to_csv(csv_file_path, metrics, organ_feature, epoch, epoch_time):
    """Appends training and validation metrics to a CSV file."""
    with open(csv_file_path, mode='a', newline='') as file:
        writer = csv.writer(file)
        train_precision = metrics['train']['macro avg']['precision']
        train_recall = metrics['train']['macro avg']['recall']
        train_f1 = metrics['train']['macro avg']['f1-score']
        train_loss = metrics['train_loss']
        train_acc = metrics['train_accuracy']

        val_precision = metrics['validation']['macro avg']['precision']
        val_recall = metrics['validation']['macro avg']['recall']
        val_f1 = metrics['validation']['macro avg']['f1-score']
        val_loss = metrics['val_loss']
        val_acc = metrics['val_accuracy']

        writer.writerow([organ_feature, epoch, train_loss, train_acc, val_loss, val_acc,
                         train_precision, train_recall, train_f1, val_precision, val_recall,
                         val_f1, f"{epoch_time:.2f}"])

from concurrent.futures import ThreadPoolExecutor, as_completed

# def validate_batch(data, target, num_classes, batch_idx):
#     """Validates a single batch for NaN values, image shape, and label correctness."""
#     if torch.isnan(data).any():
#         raise ValueError(f"NaN detected in data at batch {batch_idx}")
#     if torch.isnan(target).any():
#         raise ValueError(f"NaN detected in target at batch {batch_idx}")
#     if data.shape[1:] != (3, 1920, 1920):  # Check if images have expected shape
#         raise ValueError(f"Unexpected image shape at batch {batch_idx}: {data.shape}")
#     if target.max() >= num_classes or target.min() < 0:  # Ensure labels are within expected range
#         raise ValueError(f"Invalid label at batch {batch_idx}: {target}")
#     return f"Batch {batch_idx} passed"

# def validate_data_integrity(loader, num_classes):
#     """Validates data integrity by checking for NaN values, image shape consistency, and label correctness in parallel."""
#     futures = []
#     with ThreadPoolExecutor(max_workers=16) as executor:
#         for batch_idx, (data, target) in tqdm(enumerate(loader), total=len(loader), desc="Validating Data Integrity"):
#             futures.append(executor.submit(validate_batch, data, target, num_classes, batch_idx))

#         # Collect results and check for exceptions
#         for future in as_completed(futures):
#             try:
#                 result = future.result()
#                 if verbose:
#                     print(result)
#             except Exception as exc:
#                 print(f"Validation error: {exc}")
#                 raise exc


def run_training(organ_feature, project, dataset_dir, csv_file_path):
    """Trains a Swin Transformer model on a specific organ feature and logs metrics."""
    # Set device to CUDA if available
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    organ_feature_path = os.path.join(dataset_dir, organ_feature)

    if not os.path.isdir(organ_feature_path):
        print(f"Directory {organ_feature_path} does not exist. Skipping.")
        return

    clear_memory()

    # Initialize Comet.ml experiment
    experiment = comet_ml.Experiment(project_name=organ_feature)
    experiment.set_name(f"SwinTransformer_{organ_feature}")

    # Determine the number of classes
    train_dir = os.path.join(organ_feature_path, 'train')
    classes = [d.name for d in os.scandir(train_dir) if d.is_dir()]
    num_classes = len(classes)

    if num_classes < 2:
        print(f"Feature '{organ_feature}' has less than 2 classes ({num_classes}). Skipping classification.")
        experiment.end()
        return

    print(f"Feature: {organ_feature}, Number of classes: {num_classes}, Classes: {classes}")

    # Define Swin Transformer configuration for high-resolution images
    config = SwinConfig(
      img_size=1920,  # Keep high-resolution image handling
      patch_size=4,  # Smaller patch size for finer details
      in_chans=3,  # For RGB images
      num_classes=num_classes, 
      embed_dim=32,  # Reduced embedding dimension
      depths=[2, 2, 6, 2],  # Shallower model, especially in deeper stages
      num_heads=[2, 4, 8, 8],  # Fewer attention heads
      window_size=7,  # Standard window size
      mlp_ratio=0.8,  # Smaller MLP ratio to lighten the feedforward network
      qkv_bias=True,  # Keep bias for learning fine distinctions
      qk_scale=None,  # Let the model automatically scale
      drop_rate=0.1,  # Dropout for regularization to avoid overfitting
      attn_drop_rate=0.1,  # Dropout within attention layers
      drop_path_rate=0.1,  # Drop path for regularization
      norm_layer=nn.LayerNorm,  # Use LayerNorm as standard
      ape=False,  # No absolute positional encoding (can be added if you need it)
      patch_norm=True,  # Patch normalization
      use_checkpoint=False  # Avoid using checkpoints to keep the model light
  )

    # Initialize the model from scratch (random initialization)
    model = SwinForImageClassification(config)
    model.to(device)

    # Define image transformations with high-resolution handling
    transform = transforms.Compose([
        transforms.RandomRotation(15),  # Rotation at 15 degrees
        transforms.RandomResizedCrop((1920, 1920), scale=(0.5, 1.0)),  # Crop to 1920x1920
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406],  # Standard ImageNet means
                             std=[0.229, 0.224, 0.225])   # Standard ImageNet stds
    ])

    # Load datasets
    train_dataset = datasets.ImageFolder(train_dir, transform=transform)
    val_dir = os.path.join(organ_feature_path, 'val')
    val_dataset = datasets.ImageFolder(val_dir, transform=transform)

    # # Check that the labels are within the expected range
    # try:
    #     check_label_range(train_dataset, num_classes)
    #     check_label_range(val_dataset, num_classes)
    #     print("Label range check passed.")
    # except ValueError as e:
    #     print(f"Label range error: {e}")
    #     experiment.end()
    #     return

    # Compute class weights for imbalanced datasets
    targets = np.array(train_dataset.targets)
    class_sample_counts = np.bincount(targets)

    # Invert class sample counts to get weights for each class
    class_weights = 1.0 / class_sample_counts
    samples_weights = class_weights[targets]

    # Create a WeightedRandomSampler
    sampler = WeightedRandomSampler(samples_weights, len(samples_weights))

    class_weights_tensor = torch.FloatTensor(class_weights).to(device)
    print(f"Class weights for {organ_feature}: {class_weights}")

    # Define Focal Loss function (replacing CrossEntropyLoss)
    criterion = FocalLoss(alpha=0.25, gamma=2.0)

    # Define optimizer and scheduler
    optimizer = optim.AdamW(model.parameters(), lr=1e-4)  # Initial learning rate
    scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=1, gamma=0.9)  # Aggressive decay

    num_epochs = 50
    best_val_loss = float('inf')
    patience = 5
    epochs_no_improve = 0

    # Initialize GradScaler for mixed precision
    scaler = torch.amp.GradScaler()

    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')

    # Define save directory
    save_dir = os.path.join(trainings_dir, organ_feature)
    os.makedirs(save_dir, exist_ok=True)
    model_save_path = os.path.join(save_dir, f'best_{organ_feature}_{num_classes}_classes_{timestamp}.pth')

    # Validate data integrity before training
    train_loader = DataLoader(train_dataset, num_workers=max_workers, prefetch_factor=prefetch_factor,  batch_size=batch_size, sampler=sampler)
    val_loader = DataLoader(val_dataset, num_workers=max_workers,  prefetch_factor=prefetch_factor, batch_size=batch_size, shuffle=False)

    # try:
    #     validate_data_integrity(train_loader, num_classes)
    #     validate_data_integrity(val_loader, num_classes)
    #     print("Data integrity check passed.")
    # except ValueError as e:
    #     print(f"Data integrity error: {e}")
    #     experiment.end()
    #     return

    for epoch in range(num_epochs):
        epoch_start_time = time.time()  # Start time for epoch

        model.train()
        train_loss = 0.0
        train_correct = 0
        train_total = 0
        all_labels = []
        all_preds = []

        for images, labels in tqdm(train_loader, desc=f"Training Epoch {epoch+1}/{num_epochs}"):
            images = images.to(device, non_blocking=True)
            labels = labels.to(device, non_blocking=True)

            optimizer.zero_grad()
            with torch.amp.autocast(device_type='cuda', dtype=torch.float16):
                outputs = model(images).logits
                loss = criterion(outputs, labels)

            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()

            train_loss += loss.item() * images.size(0)
            _, predicted = torch.max(outputs.data, 1)
            train_total += labels.size(0)
            train_correct += (predicted == labels).sum().item()

            all_labels.extend(labels.cpu().numpy())
            all_preds.extend(predicted.cpu().numpy())

        train_loss /= train_total
        train_accuracy = train_correct / train_total

        # Validation
        model.eval()
        val_loss = 0.0
        val_correct = 0
        val_total = 0
        all_val_labels = []
        all_val_preds = []

        with torch.no_grad():
            for images, labels in tqdm(val_loader, desc="Validation"):
                images = images.to(device, non_blocking=True)
                labels = labels.to(device, non_blocking=True)

                with torch.amp.autocast(device_type='cuda', dtype=torch.float16):
                    outputs = model(images).logits
                    loss = criterion(outputs, labels)

                val_loss += loss.item() * images.size(0)
                _, predicted = torch.max(outputs.data, 1)
                val_total += labels.size(0)
                val_correct += (predicted == labels).sum().item()

                all_val_labels.extend(labels.cpu().numpy())
                all_val_preds.extend(predicted.cpu().numpy())

        val_loss /= val_total
        val_accuracy = val_correct / val_total

        # Metrics Calculation (Precision, Recall, F1-score)
        train_metrics = classification_report(all_labels, all_preds, target_names=classes, output_dict=True)
        val_metrics = classification_report(all_val_labels, all_val_preds, target_names=classes, output_dict=True)

        metrics = {
            'train_loss': train_loss,
            'train_accuracy': train_accuracy,
            'val_loss': val_loss,
            'val_accuracy': val_accuracy,
            'train': train_metrics,
            'validation': val_metrics
        }

        epoch_end_time = time.time()
        epoch_time_in_seconds = epoch_end_time - epoch_start_time
        epoch_time_in_hours = epoch_time_in_seconds / 3600

        append_metrics_to_csv(csv_file_path, metrics, organ_feature, epoch + 1, epoch_time_in_hours)

        # Adjust learning rate
        scheduler.step()

        if val_loss < best_val_loss:
            best_val_loss = val_loss
            epochs_no_improve = 0
            torch.save(model.state_dict(), model_save_path)
            if verbose:
                print(f"Epoch {epoch+1}: Validation loss improved. Model saved.")
        else:
            epochs_no_improve += 1
            if verbose:
                print(f"Epoch {epoch+1}: No improvement in validation loss.")
            if epochs_no_improve >= patience:
                print("Early stopping triggered.")
                break

        clear_memory()

    metadata_dict = {
        "description": f"Model for {organ_feature} trained on {num_classes} classes",
        "classes": classes
    }
    metadata_save_path = os.path.join(save_dir, f'metadata_{organ_feature}_{timestamp}.json')
    with open(metadata_save_path, 'w') as f:
        json.dump(metadata_dict, f, indent=4)

    experiment.end()

def main():
    csv_file_path = os.path.join(trainings_dir, 'trainings_metrics.csv')

    print(csv_file_path)
    if not os.path.exists(csv_file_path):
        os.makedirs(csv_file_path, exist_ok=True)
        with open(csv_file_path, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(['organ_feature', 'epoch', 'train_loss', 'train_accuracy', 'val_loss', 'val_accuracy',
                             'train_precision', 'train_recall', 'train_f1', 'val_precision', 'val_recall',
                             'val_f1', 'epoch_time(hours)'])

    for organ_feature in os.listdir(dataset_dir):
        organ_feature_path = os.path.join(dataset_dir, organ_feature)
        if not os.path.isdir(organ_feature_path):
            continue  # Skip if not a directory
        try:
            run_training(organ_feature, project, dataset_dir, csv_file_path)
            print(f"Training completed for {organ_feature}.")
        except Exception as e:
            print(f"Error with {organ_feature}: {e}")
            time.sleep(1)

if __name__ == "__main__":
    main()`
